demo_modules/chicago-brick-live/chicago-brick-live.js

```diff
@@ -0,0 +1,264 @@
+
+    // Basic client info.
+    this.client = {
+          x: this.surface.virtualOffset.x,
+          y: this.surface.virtualOffset.y,
+    };
+
+    // Use the default code until told otherwise.
+    this.setClientCode({ code: "", controlled: false });
+
+    // The event we listen to for new code.
+    this.newCodeEvent = `code(${getClientKey(this.client)})`;
+    network.on(this.newCodeEvent, this.setClientCode.bind(this));
+
+    // Ask for some code to run.
+    network.emit('requestCode', { client: this.client });
+  }
+
+  extendCanvas(canvas) {
+    canvas._imageCache = {};
+
+    canvas.writeText = function(x, y, text, style, font, extraProps) {
+      _.extend(this, {textAlign: "left"}, extraProps);
+
+      this.font = font || "120px Arial";
+      this.fillStyle = style || "white";
+      this.baseBaseline = "bottom";
+      this.fillText(text, x, y);
+    };
+
+    canvas.draw = {
+      image: function(x, y, url, scale) {
+        scale = scale || 1;
+
+        // Only load each image once since everything is done in the draw loop.
+        let image = canvas._imageCache[url];
+
+        if (!image) {
+          image = new Image();
+          image.src = url;
+          canvas._imageCache[url] = image;
+        }
+
+        canvas.drawImage(image, x, y, scale * image.width, scale * image.height);
+      },
+      rectangle: function(rect, style) {
+        canvas.fillStyle = style || "white";
+        canvas.fillRect(rect.left || rect.x, rect.top || rect.y,
+                        rect.width || rect.w, rect.height|| rect.h);
+      },
+      circle: function(x, y, radius, style) {
+        canvas.fillStyle = style || "white";
+        canvas.beginPath();
+        canvas.arc(x, y, radius, 0, 2 * Math.PI);
+        canvas.fill();
+      }
+    };
+
+    return canvas;
+  }
+
+  setClientCode(clientCode) {
+    // Merge the new client code with the existing.
+    this.clientCode = _.extend(this.clientCode || {}, clientCode, { time0: undefined });
+
+    try {
+      // Draw params.
+      this.clientCode.drawParams = {
+        canvas: this.canvas,
+        time: undefined,
+        globalTime: undefined,
+        screen: this.screen,
+      };
+      const params = _.keys(this.clientCode.drawParams);
+      const params_and_code = params.concat([this.clientCode.code]);
+      this.clientCode.draw = new Function(...params_and_code);
+    } catch (e) {
+      // If there is a syntax error "new Function" will fail, replace code with
+      // error message.
+      this.setClientCode({ code: ClientCodeError(e.message) });
+    }
+  }
+
+  draw(time, delta) {
+    this.canvas.draw.rectangle(this.screen, 'black');
+
+    this.clientCode.time0 = this.clientCode.time0 || time;
+
+    try {
+      var params = Object.assign({}, this.clientCode.drawParams);
+      params.time = time - this.clientCode.time0;
+      params.globaltime = time;
+
+      this.clientCode.draw(..._.values(params));
```

Comment:
The ordering of the values of an Object is not necessarily stable over multiple calls if the Object is changed. Is there a better way to store this structure?